File: leetcode/88_Merge_Sorted_Array.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        
        if (n == 1 && m == 0) {
            nums1[0] = nums2[0];
            return;
        }
        
        int k = m + n - 1;
        // int count = m - 1;

        for (int i=n-1; i>=0; i--) {
            cout << "i = " << i << endl;
            while (m > 0) {
                if (nums2[i] >= nums1[m-1]) {
                    nums1[k--] = nums2[i];
                    break;
                } else {
                    nums1[k--] = nums1[m-1];
                    m--;
                }
            }
            if (m == 0) {
                nums1[k--] = nums2[i];
            }
        }
    }
};

int main(int argc, char* argv[]) {

    // vector<int> nums1 = {2,0};
    // vector<int> nums2 = {1};

    vector<int> nums1 = {1,2,3,0,0,0};
    vector<int> nums2 = {1,5,6};

    // vector<int> nums1 = {4,5,6,0,0,0};
    // vector<int> nums2 = {1,2,3};

    Solution s;
    s.merge(nums1, 3, nums2, 3);

    for (auto num1 : nums1) {
        cout << num1 << " ";
    }
    cout << endl;

    return 0;
}

File: leetcode/88_Merge_Sorted_Array.rs
struct Solution {}

impl Solution {
    pub fn merge(nums1: &mut Vec<i32>, m: i32, nums2: &mut Vec<i32>, n: i32) {

        if n == 0 {
            return
        }

        let mut i = 0;
        if m == 0 {
            while i < n {
                nums1[i as usize] = nums2[i as usize];
                i += 1;
            }            
        }

        let mut i = m - 1;
        let mut j = n - 1;
        let mut k = m + n - 1;

        while i >= 0 && j >= 0 {
            if nums1[i as usize] > nums2[j as usize] {
                nums1[k as usize] = nums1[i as usize];
                i -= 1;
            } else {
                nums1[k as usize] = nums2[j as usize];
                j -= 1;
            }

            k -= 1;
        }

        while j >= 0 {
            nums1[k as usize] = nums2[j as usize];
            k -= 1;
            j -= 1;
        }
    }
}

fn main() {

    Solution::merge(&mut vec![1,2,3,0,0,0], 3, &mut vec![2,5,6], 3);
}

